// eslint-plugin-sonarjs/cjs/S6564/generated-meta.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6564/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.sonarKey = exports.meta = void 0;
// DO NOT EDIT! This file is autogenerated by "npm run generate-meta"
exports.meta = {
    type: 'suggestion',
    docs: {
        description: 'Redundant type aliases should not be used',
        recommended: true,
        url: 'https://sonarsource.github.io/rspec/#/rspec/S6564/javascript',
        requiresTypeChecking: false,
    },
    fixable: undefined,
    deprecated: false,
};
exports.sonarKey = 'S6564';
