// eslint-plugin-sonarjs/cjs/S6564/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6564/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const COMMON_NODE_TYPES = new Set([
    'TSAnyKeyword',
    'TSBigIntKeyword',
    'TSBooleanKeyword',
    'TSNeverKeyword',
    'TSNullKeyword',
    'TSNumberKeyword',
    'TSStringKeyword',
    'TSSymbolKeyword',
    'TSUndefinedKeyword',
    'TSUnknownKeyword',
    'TSVoidKeyword',
]);
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, {
        messages: {
            redundantTypeAlias: 'Remove this redundant type alias and replace its occurrences with "{{type}}".',
        },
    }),
    create(context) {
        return {
            TSTypeAliasDeclaration(node) {
                const { id, typeAnnotation } = node;
                if (COMMON_NODE_TYPES.has(typeAnnotation.type) || (0, index_js_1.isTypeAlias)(typeAnnotation, context)) {
                    const sourceCode = context.sourceCode;
                    const tpe = sourceCode.getTokens(typeAnnotation)[0];
                    context.report({
                        messageId: 'redundantTypeAlias',
                        node: id,
                        data: {
                            type: tpe.value,
                        },
                    });
                }
            },
        };
    },
};
