// eslint-plugin-sonarjs/cjs/S3981/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S3981
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const CollectionLike = ['Array', 'Map', 'Set', 'WeakMap', 'WeakSet'];
const CollectionSizeLike = ['length', 'size'];
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, {
        messages: {
            fixCollectionSizeCheck: 'Fix this expression; {{propertyName}} of "{{objectName}}" is always greater or equal to zero.',
            suggestFixedSizeCheck: 'Use "{{operator}}" for {{operation}} check',
        },
        hasSuggestions: true,
    }),
    create(context) {
        const services = context.sourceCode.parserServices;
        const isTypeCheckerAvailable = (0, index_js_1.isRequiredParserServices)(services);
        return {
            BinaryExpression: (node) => {
                if (['<', '>='].includes(node.operator)) {
                    const lhs = node.left;
                    const rhs = node.right;
                    if (isZeroLiteral(rhs) && lhs.type === 'MemberExpression') {
                        const { object, property } = lhs;
                        if (property.type === 'Identifier' &&
                            CollectionSizeLike.includes(property.name) &&
                            (!isTypeCheckerAvailable || isCollection(object, services))) {
                            context.report({
                                messageId: 'fixCollectionSizeCheck',
                                data: {
                                    propertyName: property.name,
                                    objectName: context.sourceCode.getText(object),
                                },
                                node,
                                suggest: getSuggestion(node, property.name, context),
                            });
                        }
                    }
                }
            },
        };
    },
};
function isZeroLiteral(node) {
    return node.type === 'Literal' && node.value === 0;
}
function isCollection(node, services) {
    const checker = services.program.getTypeChecker();
    const tp = checker.getTypeAtLocation(services.esTreeNodeToTSNodeMap.get(node));
    return !!tp.symbol && CollectionLike.includes(tp.symbol.name);
}
function getSuggestion(expr, operation, context) {
    const { left, operator } = expr;
    const operatorToken = context.sourceCode.getTokenAfter(left, token => token.value === operator);
    const fixedOperator = operator === '<' ? '==' : '>';
    return [
        {
            messageId: 'suggestFixedSizeCheck',
            data: {
                operation,
                operator: fixedOperator,
            },
            fix: fixer => fixer.replaceText(operatorToken, fixedOperator),
        },
    ];
}
