// eslint-plugin-sonarjs/cjs/S5869/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S5869/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const rule_template_js_1 = require("../helpers/regex/rule-template.js");
const simplified_regex_character_class_js_1 = require("../helpers/regex/simplified-regex-character-class.js");
const location_js_1 = require("../helpers/regex/location.js");
exports.rule = (0, rule_template_js_1.createRegExpRule)(context => {
    let flags;
    return {
        onRegExpLiteralEnter: (node) => {
            flags = node.flags;
        },
        onCharacterClassEnter: (node) => {
            const duplicates = new Set();
            const characterClass = new simplified_regex_character_class_js_1.SimplifiedRegexCharacterClass(flags);
            node.elements.forEach(element => {
                const intersections = new simplified_regex_character_class_js_1.SimplifiedRegexCharacterClass(flags, element).findIntersections(characterClass);
                if (intersections.length > 0) {
                    intersections.forEach(intersection => duplicates.add(intersection));
                    duplicates.add(element);
                }
                characterClass.add(element);
            });
            if (duplicates.size > 0) {
                const [primary, ...secondaries] = duplicates;
                const secondaryLocations = [];
                for (const secondary of secondaries) {
                    const loc = (0, location_js_1.getRegexpLocation)(context.node, secondary, context);
                    if (loc) {
                        secondaryLocations.push((0, index_js_1.toSecondaryLocation)({ loc }, 'Additional duplicate'));
                    }
                }
                context.reportRegExpNode({
                    message: 'Remove duplicates in this character class.',
                    node: context.node,
                    regexpNode: primary,
                }, secondaryLocations);
            }
        },
    };
}, (0, index_js_1.generateMeta)(meta_js_1.meta, undefined, true));
