// eslint-plugin-sonarjs/cjs/S6080/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6080/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const MESSAGE = 'Set this timeout to 0 if you want to disable it, otherwise use a value lower than 2147483648.';
const MAX_DELAY_VALUE = 2_147_483_647;
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta),
    create(context) {
        if (!index_js_1.Chai.isImported(context)) {
            return {};
        }
        const constructs = [];
        return {
            CallExpression: (node) => {
                if (index_js_1.Mocha.isTestConstruct(node)) {
                    constructs.push(node);
                    return;
                }
                if (constructs.length > 0) {
                    checkTimeoutDisabling(node, context);
                }
            },
            'CallExpression:exit': (node) => {
                if (index_js_1.Mocha.isTestConstruct(node)) {
                    constructs.pop();
                }
            },
        };
    },
};
function checkTimeoutDisabling(node, context) {
    if ((0, index_js_1.isMethodCall)(node) && node.arguments.length > 0) {
        const { callee: { object, property }, arguments: [value], } = node;
        if ((0, index_js_1.isThisExpression)(object) &&
            (0, index_js_1.isIdentifier)(property, 'timeout') &&
            isDisablingTimeout(value, context)) {
            context.report({
                message: MESSAGE,
                node: value,
            });
        }
    }
}
function isDisablingTimeout(timeout, context) {
    const usage = (0, index_js_1.getUniqueWriteUsageOrNode)(context, timeout);
    return (0, index_js_1.isNumberLiteral)(usage) && usage.value > MAX_DELAY_VALUE;
}
