// eslint-plugin-sonarjs/cjs/S6302/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6302/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const result_js_1 = require("../helpers/result.js");
const iam_js_1 = require("../helpers/aws/iam.js");
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const MESSAGES = {
    message: 'Make sure granting all privileges is safe here.',
    secondary: 'Related effect',
};
exports.rule = (0, iam_js_1.AwsIamPolicyTemplate)(allPrivilegesStatementChecker, (0, index_js_1.generateMeta)(meta_js_1.meta, undefined, true));
function allPrivilegesStatementChecker(expr, ctx, options) {
    const properties = (0, result_js_1.getResultOfExpression)(ctx, expr);
    const effect = (0, iam_js_1.getSensitiveEffect)(properties, ctx, options);
    const action = getSensitiveAction(properties, options);
    if (effect.isMissing && action) {
        (0, index_js_1.report)(ctx, {
            message: MESSAGES.message,
            node: action,
        });
    }
    else if (effect.isFound && action) {
        (0, index_js_1.report)(ctx, {
            message: MESSAGES.message,
            node: action,
        }, [(0, index_js_1.toSecondaryLocation)(effect.node, MESSAGES.secondary)]);
    }
}
function getSensitiveAction(properties, options) {
    return getActionLiterals(properties, options).find(iam_js_1.isAnyLiteral);
}
function getActionLiterals(properties, options) {
    return properties.getProperty(options.actions.property).asStringLiterals();
}
