// eslint-plugin-sonarjs/cjs/S1264/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S1264
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, {
        messages: {
            replaceForWithWhileLoop: 'Replace this "for" loop with a "while" loop.',
        },
        fixable: 'code',
    }),
    create(context) {
        return {
            ForStatement(forLoop) {
                const forKeyword = context.sourceCode.getFirstToken(forLoop);
                if (!forLoop.init && !forLoop.update && forLoop.test && forKeyword) {
                    context.report({
                        messageId: `replaceForWithWhileLoop`,
                        loc: forKeyword.loc,
                        fix: getFix(forLoop),
                    });
                }
            },
        };
        function getFix(forLoop) {
            const forLoopRange = forLoop.range;
            const closingParenthesisToken = context.sourceCode.getTokenBefore(forLoop.body);
            const condition = forLoop.test;
            if (condition && forLoopRange && closingParenthesisToken) {
                return (fixer) => {
                    const start = forLoopRange[0];
                    const end = closingParenthesisToken.range[1];
                    const conditionText = context.sourceCode.getText(condition);
                    return fixer.replaceTextRange([start, end], `while (${conditionText})`);
                };
            }
            return undefined;
        }
    },
};
