// eslint-plugin-sonarjs/cjs/S2486/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S2486/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, {
        messages: {
            handleException: "Handle this exception or don't catch it at all.",
        },
    }),
    create(context) {
        return {
            'CatchClause[param.type="Identifier"]'(node) {
                const param = node.param;
                const scope = context.sourceCode.getScope(node);
                const variable = (0, index_js_1.getVariableFromScope)(scope, param.name);
                if (variable?.references.length === 0) {
                    context.report({
                        messageId: 'handleException',
                        node,
                    });
                }
            },
        };
    },
};
