// eslint-plugin-sonarjs/cjs/S6252/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6252/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const VERSIONED_KEY = 'versioned';
const messages = {
    unversioned: 'Make sure using unversioned S3 bucket is safe here.',
    omitted: 'Omitting the "versioned" argument disables S3 bucket versioning. Make sure it is safe here.',
    secondary: 'Propagated setting',
};
exports.rule = (0, index_js_1.S3BucketTemplate)((bucketConstructor, context) => {
    const versionedProperty = (0, index_js_1.getBucketProperty)(context, bucketConstructor, VERSIONED_KEY);
    if (versionedProperty == null) {
        (0, index_js_1.report)(context, {
            message: messages.omitted,
            node: bucketConstructor.callee,
        });
        return;
    }
    const propertyLiteralValue = (0, index_js_1.getValueOfExpression)(context, versionedProperty.value, 'Literal');
    if (propertyLiteralValue?.value === false) {
        const secondaries = [];
        const isPropagatedProperty = versionedProperty.value !== propertyLiteralValue;
        if (isPropagatedProperty) {
            secondaries.push((0, index_js_1.toSecondaryLocation)((0, index_js_1.getNodeParent)(propertyLiteralValue), messages.secondary));
        }
        (0, index_js_1.report)(context, {
            message: messages.unversioned,
            node: versionedProperty,
        }, secondaries);
    }
}, (0, index_js_1.generateMeta)(meta_js_1.meta, undefined, true));
