// eslint-plugin-sonarjs/cjs/S2068/generated-meta.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S2068/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = exports.sonarKey = exports.meta = void 0;
// DO NOT EDIT! This file is autogenerated by "npm run generate-meta"
exports.meta = {
    type: 'problem',
    docs: {
        description: 'Hard-coded passwords are security-sensitive',
        recommended: true,
        url: 'https://sonarsource.github.io/rspec/#/rspec/S2068/javascript',
        requiresTypeChecking: false,
    },
    fixable: undefined,
    deprecated: false,
};
exports.sonarKey = 'S2068';
exports.schema = {
    type: 'array',
    minItems: 0,
    maxItems: 1,
    items: [
        {
            type: 'object',
            properties: {
                passwordWords: {
                    type: 'array',
                    items: {
                        type: 'string',
                    },
                },
            },
            additionalProperties: false,
        },
    ],
};
