// eslint-plugin-sonarjs/cjs/S4622/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S4622/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const DEFAULT_THRESHOLD = 3;
const messages = {
    refactorUnion: 'Refactor this union type to have less than {{threshold}} elements.',
};
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, { messages, schema: meta_js_1.schema }),
    create(context) {
        return {
            TSUnionType: (node) => {
                const union = node;
                if (isUsedWithUtilityType(union)) {
                    return;
                }
                const threshold = context.options[0]?.threshold ?? DEFAULT_THRESHOLD;
                if (union.types.length > threshold && !isFromTypeStatement(union)) {
                    context.report({
                        messageId: 'refactorUnion',
                        data: {
                            threshold: `${threshold}`,
                        },
                        node,
                    });
                }
            },
        };
    },
};
function isFromTypeStatement(node) {
    return node.parent.type === 'TSTypeAliasDeclaration';
}
function isUsedWithUtilityType(node) {
    return (node.parent.type === 'TSTypeParameterInstantiation' &&
        node.parent.parent.type === 'TSTypeReference' &&
        (0, index_js_1.isIdentifier)(node.parent.parent.typeName, ...index_js_1.UTILITY_TYPES));
}
