// eslint-plugin-sonarjs/cjs/S6627/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S6627/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const messages = {
    default: 'Do not use internal APIs of your dependencies',
};
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, { messages }),
    create(context) {
        return {
            CallExpression(node) {
                if ((0, index_js_1.isRequire)(node)) {
                    const [arg] = node.arguments;
                    if ((0, index_js_1.isStringLiteral)(arg) && arg.value.includes('node_modules')) {
                        context.report({
                            node,
                            messageId: 'default',
                        });
                    }
                }
            },
            ImportDeclaration(node) {
                const moduleName = node.source.value;
                if (moduleName.includes('node_modules')) {
                    context.report({
                        node,
                        messageId: 'default',
                    });
                }
            },
        };
    },
};
