// eslint-plugin-sonarjs/cjs/S7059/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S7059/javascript
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta_js_1 = require("./meta.js");
const flaggedStatements = new Set();
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta_js_1.meta, {
        messages: {
            noAsyncConstructor: 'Refactor this asynchronous operation outside of the constructor.',
        },
    }),
    create(context) {
        const services = context.sourceCode.parserServices;
        if (!(0, index_js_1.isRequiredParserServices)(services)) {
            return {};
        }
        /**
         * Given a Promise call, get the parent statement of the async call.
         * We want to ensure that it is inside a constructor, but not part of a function declaration:
         * constructor() {
         *  foo();
         * }
         * and not
         * constructor() {
         *  myFunction = () => { foo() }
         * }
         * @param node : promise call
         */
        function asyncStatementInsideConstructor(node) {
            let classConstructor;
            let statement;
            context.sourceCode.getAncestors(node).forEach(node => {
                if (node.type === 'MethodDefinition' && node.kind === 'constructor') {
                    classConstructor = node;
                }
                if (classConstructor && node.type.endsWith('Statement')) {
                    statement = node;
                }
                // If we find a function declaration it should not be considered as part of the constructor
                if (classConstructor && statement && (0, index_js_1.isFunctionNode)(node)) {
                    statement = undefined;
                    classConstructor = undefined;
                }
            });
            return statement;
        }
        return {
            CallExpression(node) {
                if (!(0, index_js_1.isThenable)(node, services)) {
                    return;
                }
                // we want to raise on the parent statement
                const statement = asyncStatementInsideConstructor(node);
                if (statement && !flaggedStatements.has(statement)) {
                    flaggedStatements.add(statement);
                    context.report({
                        node: statement,
                        messageId: 'noAsyncConstructor',
                    });
                }
            },
            'Program:exit'() {
                flaggedStatements.clear();
            },
        };
    },
};
